### ISSUE TITLE
Help voor package developers?

### ISSUE THREAD

**Issue by jhubers** [DynamoDS/Dynamo#6956]
When you want to start a new issue it shows:
_If this issue is **not** a bug report or improvement request, please check the [Dynamo forum](https://forum.dynamobim.com/), and start a thread there to discuss your issue._ But that forum seems to have no support of Dynamo programmers.

Does this mean that there is no help anymore for package developers? 

My question is: How to subscribe to an event when the home workspace is closing, and if certain condition is not true how to cancel the closing. In this case I have users type some information in a DataGrid and want to warn, that they lose this information.

**Comment by sharadkjaiswal** [comment]
@jhubers You can get the notifications by registering your event handlers to following events 
```
WorkspaceEvents.WorkspaceRemoveStarted 
WorkspaceEvents.WorkspaceRemoved
```
But you can't cancel the closing at this moment.

**Comment by jhubers** [comment]
@sharadkjaiswal thanks. And the main question? Is there still help for amateur developers like me? And I was wondering if I could subscribe to model.CleaningUp and then somehow set DynamoViewModel.CloseHomeWorkspaceCommand.CanExecute = false ? Would that also cancel e.g. shutdown of Dynamo or opening other Home Workspace or a new one? Should I otherwise make a request for this?

**Comment by sharadkjaiswal** [comment]
CanExecute is checked before executing the command, it won't have any impact when the command is already executing.

**Comment by sharadkjaiswal** [comment]
Can you please explain the scenario in detail? When the workspace is closed, it will prompt users to save the workspace. Are implementing a UI node and you are wondering how to serialize the user input data?

**Comment by jhubers** [comment]
Not quit. I have a custom XAML control window with a DataGrid that shows the content of a csv file where others can also write and read. To this DataGrid is added some information from Dynamo nodes that the user wants to share. E.g. the x,y,z values of a point where others need to connect there objects. This information of course is serialized. But in the DataGrid is also a column for comments, that is much handier to edit directly. This information is not in the *.dyn file. So you can not serialize it. When another home workspace is opened or Dynamo is shut down, you lose the edits to the comments.

**Comment by sharadkjaiswal** [comment]
If the comments in DataGrid is not serialized then how do you restore it across different sessions? Why don't you serialize the comments in DYN file?

**Comment by jhubers** [comment]
It is stored in the shared csv file. The comments can be edited by all team members. I don't know a way to get the edits into the DYN file. In earlier versions I solved it by always adding the comments after a slash, then it could be serialized as a string input to a node. But comments tend to become very long. I could maybe create a node that outputs the comment of an item. Take that as an input to another node which permits the editing and then use that as input for again another node and serialize that node. But it is not user-friendly. Too many nodes. Editing inside the displayed DataGrid is better.

**Comment by sharadkjaiswal** [comment]
Sounds like you are doing something interesting.
From your node instance you can register event handler for workspace remove started and there you can save the comments to related csv file.

**Comment by jhubers** [comment]
Yes, but it could be that the user was not finished editing the comment. Saving to the csv file gives an alert through filesystemwatcher and other team members could get unfinished edited comments...

**Comment by jhubers** [comment]
@sharadkjaiswal BTW the events you mentioned fire when you close a workspace that was used for Create Node From Selection. When opening a new Home Workspace or opening an existing DYN file Dynamo.Events.WorkspaceEvents.WorkspaceClearing is the event that fires. And DynamoModel.CleaningUp fires when Dynamo shuts down. I hope that in next release there will be a way to cancel the clearing of a workspace and then also the shutting down of Dynamo.
